### ISSUE TITLE
Npgsql v5 bug causing NServiceBus Saga deserialization exceptions

### ISSUE THREAD

**Issue by greatbob6** [Particular/NServiceBus.Persistence.Sql#584]
I realize this isn't a NServiceBus specific issue but thought I'd share in case you guys weren't aware of it.  While working on implementing a saga using PostgreSQL persistence for a company project, I started getting errors during deserialization of the SagaData object and was able to track it down to a bug in Npgsql.  In some cases, Npgsql was not actually reading all of the data out of the Data column and thus the deserialization was trying to work with incorrect data.

Apparently, someone else found this as well and a fix was merged in a few days ago:

https://github.com/npgsql/npgsql/commit/fb1204007da3edfd845e8db034bbb93db4401a95

According to the tagging, that fix should be deployed in Npgsql 5.0.2, so this issue applies to Npgsql version >= 5.0.0 and < 5.0.2.

For now, I will go back to the previous major version of Npgsql and hopefully that will work properly.

**Comment by ramonsmits** [comment]
@greatbob6 thanks for sharing your findings. Would you know what the exception was that you got? For SEO purposes it would be great information to have in this issue.

**Comment by greatbob6** [comment]
We were seeing the following exception(s) thrown in our application, with the first exception happening the first time we encountered the issue after restarting the app, and the second exception happening on subsequent processing until restarting:

```
Npgsql.NpgsqlOperationInProgressException

  HResult=0x80004005
  Message=A command is already in progress: 
select
    "Id",
    "SagaTypeVersion",
    "Concurrency",
    "Metadata",
    "Data"
from "public"."Eda_Standard_Webjob_BillStandardizationSaga"
where "Id" = @Id
for update

  Source=Npgsql
  StackTrace:
   at Npgsql.NpgsqlConnector.<StartUserAction>g__DoStartUserAction|233_0(<>c__DisplayClass233_0& )
   at Npgsql.NpgsqlConnector.StartUserAction(ConnectorState newState, NpgsqlCommand command, CancellationToken cancellationToken, Boolean attemptPgCancellation)
   at Npgsql.NpgsqlConnector.StartUserAction(CancellationToken cancellationToken, Boolean attemptPgCancellation)
   at Npgsql.NpgsqlConnector.<Rollback>d__197.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Npgsql.NpgsqlTransaction.Rollback()
   at Npgsql.NpgsqlTransaction.Dispose(Boolean disposing)
   at System.Data.Common.DbTransaction.Dispose()
   at StorageSession.Dispose() in /_/src/SqlPersistence/SynchronizedStorage/StorageSession.cs:line 53
   at NServiceBus.LoadHandlersConnector.<Invoke>d__1.MoveNext() in /_/src/NServiceBus.Core/Pipeline/Incoming/LoadHandlersConnector.cs:line 58
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at CurrentSessionBehavior.<Invoke>d__2.MoveNext() in /_/src/SqlPersistence/SynchronizedStorage/CurrentSessionBehavior.cs:line 18
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

  This exception was originally thrown at this call stack:
    Npgsql.NpgsqlConnector.StartUserAction.__DoStartUserAction|233_0(ref Npgsql.NpgsqlConnector.<>c__DisplayClass233_0)
    Npgsql.NpgsqlConnector.StartUserAction(Npgsql.ConnectorState, Npgsql.NpgsqlCommand, System.Threading.CancellationToken, bool)
    Npgsql.NpgsqlConnector.StartUserAction(System.Threading.CancellationToken, bool)
    Npgsql.NpgsqlConnector.Rollback(bool, System.Threading.CancellationToken)
    System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
    System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
    Npgsql.NpgsqlTransaction.Rollback()
    Npgsql.NpgsqlTransaction.Dispose(bool)
    System.Data.Common.DbTransaction.Dispose()
    ...
    [Call Stack Truncated]
```



```
System.Net.Internals.SocketExceptionFactory.ExtendedSocketException
  HResult=0x80004005
  Message=A non-blocking socket operation could not be completed immediately. X.X.X.X:5432
  Source=System.Net.Sockets
  StackTrace:
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
```




And after digging into it in the debugger, we found the following was the initial exception that then bubbled up to cause the first exception above:

```
Newtonsoft.Json.JsonReaderException
  HResult=0x80131500
  Message=After parsing a value an unexpected character was encountered: :. Path 'BillDetailServiceMappings[13].ServiceName', line 1, position 3831.
  Source=Newtonsoft.Json
  StackTrace:
   at Newtonsoft.Json.JsonTextReader.ParsePostValue(Boolean ignoreComments)
```

**Comment by mikeminutillo** [comment]
@greatbob6 now that npgsql 5.0.3 has been released, have you had a chance to upgrade and confirm that it resolves this issue?

**Comment by mikeminutillo** [comment]
@greatbob6 we believe this issue has been resolved. If you are still experiencing the same issue, please don't hesitate to comment on this issue and we can look into it again.

**Comment by greatbob6** [comment]
For other reasons, we ended up moving away from Sql persistence so I've been unable to test this fix, but looking at the Npgsql code, it appears that the fix is in 5.0.2+ so it should be good.

**Comment by mikeminutillo** [comment]
@greatbob6 I'm glad to hear you're unstuck. If you're willing to share, I'd love to hear the details of why you moved away from Sql Persistence and what you selected instead. If you're not comfortable putting details here on the issue, please feel free to [send us an email](https://particular.net/contact). If you mention me, whoever picks it up will get it to me :)
